### PR TITLE
Implement HTTPS certificate options

### PR DESCRIPTION
**Pull request by fredreichbier**
This implements (1) hostname validation against a custom certificate (2) disabling hostname validation.

See #27.

### CODE REVIEW

File: deploy/ubuntu-config/proxy.ini [hunk 1]
```diff
@@ -1,6 +1,12 @@
 [privacyidea]
 # URL of your privacyIDEA installation
 instance = http://192.0.2.2
+# If you use HTTPS, you can manually pass a certificate (PEM format)
+# If this option is omitted, the presented certificate is validated against the OS certificate
+# store.
+#certificate = /etc/privacyidea/server.pem
+# You can also disable hostname validation completely, which is not recommended.
```
Comment (cornelinux) [inline deploy/ubuntu-config/proxy.ini]: It is not quite clear, that ``verify`` needs to be set to ``True`` in order for ``certificate`` to take effect.
Maybe you can add something like:

* Default of ``verify`` is True
* If you want the manually passed certificate to be verified, you need to set ``verify=True``.

File: pi_ldapproxy/proxy.py [hunk 2]
```diff
@@ -255,7 +258,18 @@ class ProxyServerFactory(protocol.ServerFactory):
     def __init__(self, config):
         # NOTE: ServerFactory.__init__ does not exist?
         # Read configuration options.
-        self.agent = Agent(reactor)
+        if config['privacyidea']['verify']:
+            if config['privacyidea']['certificate']:
+                certificate = Certificate.loadPEM(FilePath(config['privacyidea']['certificate']).getContent())
+                log.info('Pinning privacyIDEA HTTPS certificate {certificate!r}', certificate=certificate)
```
Comment (cornelinux) [inline pi_ldapproxy/proxy.py]: Does this logentry looks readable at all? The representation for the certificate? Is the filename of the certificate contained? Maybe in the logfile it is better, to know from **where** the certificate was loaded?

Comment (fredreichbier) [inline pi_ldapproxy/proxy.py]: The log entry looks like this:
```
Jul 14 17:15:23 pxproxy twistd[15354]: 2017-07-14T17:15:23+0200 [pi_ldapproxy.proxy#info] Pinning privacyIDEA HTTPS certificate <Certificate Subject=privacyideaserver Issuer=privacyideaserver>
```

... I find this quite nice, but I can also change the log message to contain the filename instead :)

Comment (cornelinux) [inline pi_ldapproxy/proxy.py]: ...or both.

File: deploy/ubuntu-config/proxy.ini (hunk 1 above)
Comment (fredreichbier) [inline deploy/ubuntu-config/proxy.ini]: I agree it's not very clear. I changed this, could you take a look?

File: deploy/ubuntu-config/proxy.ini [hunk 3]
```diff
@@ -1,6 +1,14 @@
...
 # URL of your privacyIDEA installation
 instance = http://192.0.2.2
+# In case you use HTTPS:
+# You can disable hostname validation completely, which is not recommended.
+#verify = False
+# By default, verify is set to True.
+# If verify=True, you can manually pass a HTTPS server certificate (PEM format).
+#certificate = /etc/privacyidea/server.pem
+# If verify=True but you do not explicitly pass a certficate,
+# the server certificate is validated against the OS certificate store.
```
Comment (cornelinux) [inline deploy/ubuntu-config/proxy.ini]: Totally clear now!

File: pi_ldapproxy/proxy.py (hunk 2 above)
Comment (cornelinux) [inline pi_ldapproxy/proxy.py]: WHo cares. This is fine. If you want to, add it later ;)

**Comment by cornelinux** [comment]
Oups, can not rebase and merge automagically

File: pi_ldapproxy/proxy.py (hunk 2 above)
Comment (fredreichbier) [inline pi_ldapproxy/proxy.py]: I added this :)

**Pull request merged**